pkg/sessions/redis/client.go

```diff
@@ -47,13 +49,56 @@ func newClusterClient(c *redis.ClusterClient) Client {
 }
 
 func (c *clusterClient) Get(ctx context.Context, key string) ([]byte, error) {
-	return c.WithContext(ctx).Get(key).Bytes()
+	return get(c.WithContext(ctx), key)
 }
 
 func (c *clusterClient) Set(ctx context.Context, key string, value []byte, expiration time.Duration) error {
-	return c.WithContext(ctx).Set(key, value, expiration).Err()
+	return set(c.WithContext(ctx), key, value, expiration)
 }
 
 func (c *clusterClient) Del(ctx context.Context, key string) error {
-	return c.WithContext(ctx).Del(key).Err()
+	return del(c.WithContext(ctx), key)
+}
+
+type getter interface {
+	Get(key string) *redis.StringCmd
+}
+
+var group singleflight.Group
+
+func get(cmd getter, key string) ([]byte, error) {
+	v, err, _ := group.Do(key, func() (interface{}, error) {
+		return cmd.Get(key).Bytes()
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	b, ok := v.([]byte)
+	if !ok {
+		return nil, fmt.Errorf("invalid value type: key=%s, value%v", key, v)
+	}
+	return b, nil
+}
+
+type setter interface {
+	Set(key string, value interface{}, expiration time.Duration) *redis.StatusCmd
+}
+
+func set(cmd setter, key string, value []byte, expiration time.Duration) error {
```

Comment:
Hmmm -- especially seeing `GET`, `SET` and `DEL` split into separate groups, even non-horizontal oauth2-proxy nodes could run into issues with grouped requests giving wrong data that breaks from the atomic nature of Redis that runs all commands on a single core sequentially using FIFO.

I think if we want to reduce redis trips across parallel requests -- the proper route is to use pipelining.  It still sends all the commands, but it will batch them into 1 TCP request & response.

FWIW, In the past I've found about 500 commands is the sweet spot to pipeline together before breaking them apart into the next pipeline request.